zio-http/src/main/scala/zio/http/codec/HttpCodec.scala

```diff
@@ -574,14 +574,49 @@ object HttpCodec extends ContentCodecs with HeaderCodecs with MethodCodecs with
 
     def index(index: Int): ContentStream[A] = copy(index = index)
   }
-  private[http] final case class Query[A](name: String, textCodec: TextCodec[A], index: Int = 0)
-      extends Atom[HttpCodecType.Query, A]  {
-    self =>
-    def erase: Query[Any] = self.asInstanceOf[Query[Any]]
+
+  private[http] sealed trait Query[A, I] extends Atom[HttpCodecType.Query, A] {
```

Comment:
Mono and Multi are not different structure wise. Do we need an ADT for it?
Can't we just annotate it as metadata? Or just add a boolean flag?

Reply:
@987Nabil They are different in functionality, but yes, MultiQuery could replace MonoQuery. Please check the approach: https://github.com/eshu/zio-http/pull/1. If it is ok, I can merge it and then remove monoquery.

Reply:
That one looks too complicated I think. Also it duplicates features that already exist on `HttpCodec` level like optional.

Reply:
@987Nabil Actually they have a different structures if you check the `textCodec`. Usage of it is completely different in `MonoQuery` and `MultiQuery`: for MonoQuery (which returns a scalar) it returns the result of query itself. And for `MultiQuery` the role of textCodec is different: it transforms each element of the collection.

So to merge it I need to return a collection every time, For the single element query the good candidate is `Id`, but if I drop `MonoQuery`, it requires to change all cases where scalar should be returned, so for backward compatibility it could be good to keep `MonoQuery`.

Reply:
@987Nabil Anyway I think cardinality is a good idea. What do you think?

Reply:
We do not need backwards compatible, since we are still in RC. Clean design is what we want.
I think it might be okay, to store the information if you expect single or multiple values for example for docs, but keep the multi API on the codec. Where we internally then use the codec and in user facing API, we can still behave in a differt way.
Or we try to build something like for headers, which also supports both, but has only one codec, iirc

Reply:
@987Nabil I will try.

Just curious: if you do not care about backward compatibility, why do you need Scala 2.12 support?

Reply:
@987Nabil @jdegoes 

I can't reproduce locally errors that I see in CI:
```
[error] Failed tests:
[error] 	zio.http.codec.HttpCodecSpec
[error] 	zio.http.endpoint.RoundtripSpec
[error] 	zio.http.endpoint.QueryParameterSpec
```
Could you please help?

And another problem: test `zio.http.ClientStreamingSpec` is working extremely slow (8-10 minutes) and fails even for the main branch, and this issue is not reproducible in CI.

Reply:
Also here
```
+ HttpCodecSpec
...
  + optional
    + no fallback for decoding errors - 59 ms
    - fallback for missing values
      Exception in thread "zio-fiber-30195" zio.http.codec.HttpCodecError$InvalidQueryParamCardinality: Wrong query parameter name cardinality 0, exactly one expected
      	at zio.http.codec.HttpCodecSpec.spec(HttpCodecSpec.scala:138)
      	at zio.http.codec.HttpCodecSpec.spec(HttpCodecSpec.scala:130)
```
in lines 130 and 138 of HttpCodecSpec in the local copy I see dummy tests only, they should not fail...

Reply:
Found the problem...